lib/Sema/CSSolver.cpp

```diff
@@ -269,36 +268,21 @@ bool ConstraintSystem::simplify(bool ContinueAfterFailures) {
         log << ")\n";
       }
 
-      if (solverState)
-        solverState->retireConstraint(constraint);
-
-      CG.removeConstraint(constraint);
+      retireConstraint(constraint);
       break;
 
     case SolutionKind::Solved:
-      if (solverState) {
+      if (solverState)
         ++solverState->NumSimplifiedConstraints;
-
-        // This constraint has already been solved; retire it.
-        solverState->retireConstraint(constraint);
-      }
-
-      // Remove the constraint from the constraint graph.
-      CG.removeConstraint(constraint);
+      retireConstraint(constraint);
       break;
 
     case SolutionKind::Unsolved:
       if (solverState)
         ++solverState->NumUnsimplifiedConstraints;
-
-      InactiveConstraints.push_back(constraint);
       break;
     }
 
-    // This constraint is not active. We delay this operation until
-    // after simplification to avoid re-insertion.
-    constraint->setActive(false);
```

Comment:
It's not clear to me what the comment is referring to, but it seems like this was being done later for a reason.

Are you sure we don't still need to delay this to this point?

Reply:
I think this is outdated, it's left here from SVN days when simplication was completely different, we don't check isActive() anymore besides two places in the solver, this one being one of them and another one is an assert.